src/uu/cp/src/cp.rs

```diff
@@ -1688,81 +1752,17 @@ fn copy_on_write_linux(
-    let dst_file = OpenOptions::new()
-        .write(true)
-        .truncate(true)
-        .create(true)
-        .open(dest)
-        .context(context)?;
-
-    match (reflink_mode, sparse_mode) {
-        (ReflinkMode::Always, SparseMode::Auto) => unsafe {
-            let result = libc::ioctl(dst_file.as_raw_fd(), FICLONE!(), src_file.as_raw_fd());
-
-            if result != 0 {
-                Err(format!(
-                    "failed to clone {:?} from {:?}: {}",
-                    source,
-                    dest,
-                    std::io::Error::last_os_error()
-                )
-                .into())
-            } else {
-                Ok(())
-            }
-        },
-        (ReflinkMode::Always, SparseMode::Always) | (ReflinkMode::Always, SparseMode::Never) => {
-            Err("`--reflink=always` can be used only with --sparse=auto".into())
-        }
-        (_, SparseMode::Always) => unsafe {
-            let size: usize = src_file.metadata()?.size().try_into().unwrap();
-            if libc::ftruncate(dst_file.as_raw_fd(), size.try_into().unwrap()) < 0 {
-                return Err(format!(
-                    "failed to ftruncate {:?} to size {}: {}",
-                    dest,
-                    size,
-                    std::io::Error::last_os_error()
-                )
-                .into());
-            }
-
-            let blksize = dst_file.metadata()?.blksize();
-            let mut buf: Vec<u8> = vec![0; blksize.try_into().unwrap()];
-            let mut current_offset: usize = 0;
-
-            while current_offset < size {
-                use std::io::Read;
-
-                let this_read = src_file.read(&mut buf)?;
-
-                if buf.iter().any(|&x| x != 0) {
-                    libc::pwrite(
-                        dst_file.as_raw_fd(),
-                        buf.as_ptr() as *const libc::c_void,
-                        this_read,
-                        current_offset.try_into().unwrap(),
-                    );
-                }
-                current_offset += this_read;
-            }
-            Ok(())
-        },
-        (ReflinkMode::Auto, SparseMode::Auto) | (ReflinkMode::Auto, SparseMode::Never) => unsafe {
-            let result = libc::ioctl(dst_file.as_raw_fd(), FICLONE!(), src_file.as_raw_fd());
-
-            if result != 0 {
-                fs::copy(source, dest).context(context)?;
-            }
-            Ok(())
-        },
-        (ReflinkMode::Never, _) => {
-            fs::copy(source, dest).context(context)?;
-            Ok(())
+    let result = match (reflink_mode, sparse_mode) {
+        (ReflinkMode::Never, _) => std::fs::copy(source, dest).map(|_| ()),
+        (ReflinkMode::Auto, SparseMode::Always) => sparse_copy(source, dest),
+        (ReflinkMode::Auto, _) => clone(source, dest, true),
+        (ReflinkMode::Always, SparseMode::Auto) => {
+            return Err("`--reflink=always` can be used only with --sparse=auto".into())
```

Comment:
This error says to use `--reflink=always --sparse=auto` even though that's what the user did. I think this should be:
```
(ReflinkMode::Always, SparseMode::Always | SparseMode::Never) =>
```

Or the bodies for this arm and the arm below should be switched?

Reply:
Yes, I swapped the cases. I'll fix that. Thanks.